File: src/main/parsers/tagScanner.ts
import type { LangDef } from './languages';

// Tag scanner: finds TODO/FIXME/HACK/NOTE/XXX inside line/block comments.
// It reuses language comment/string markers so plain strings do not count.

export interface FoundTag {
  kind: 'TODO' | 'FIXME' | 'HACK' | 'NOTE' | 'XXX';
  lineNo: number; // 1-based
  text: string;
}

const TAG_RE = /\b(TODO|FIXME|HACK|NOTE|XXX)\b[ \t:\-]*([^\r\n]*)/g;
const NEWLINE = /\r\n|\n|\r/;

function findStringEnd(raw: string, start: number, marker: string): number {
  let cursor = raw.indexOf(marker, start);
  while (cursor >= 0) {
    let backslashes = 0;
    for (let index = cursor - 1; index >= start && raw[index] === '\\'; index -= 1) {
      backslashes += 1;
    }
    if (backslashes % 2 === 0) return cursor;
    cursor = raw.indexOf(marker, cursor + marker.length);
  }
  return -1;
}

function pushMatches(out: FoundTag[], text: string, lineNo: number): void {
  if (!text.trim()) return;
  TAG_RE.lastIndex = 0;
  let match: RegExpExecArray | null;
  while ((match = TAG_RE.exec(text)) !== null) {
    out.push({
      kind: match[1] as FoundTag['kind'],
      lineNo,
      text: match[2].trim().slice(0, 240),
    });
  }
}

export function scanTags(content: string, lang: LangDef | null): FoundTag[] {
  const out: FoundTag[] = [];
  if (!lang) return out;

  const lines = content.split(NEWLINE);
  let inBlock: [string, string] | null = null;
  const lineMarkers = lang.line;
  const blocks = lang.block;
  const strings = lang.string ?? [];

  for (let lineIndex = 0; lineIndex < lines.length; lineIndex += 1) {
    const raw = lines[lineIndex];
    const lineNo = lineIndex + 1;
    let index = 0;
    let inString: [string, string] | null = null;
    const commentParts: string[] = [];

    while (index < raw.length) {
      if (inBlock) {
        const endIndex = raw.indexOf(inBlock[1], index);
        if (endIndex < 0) {
          commentParts.push(raw.slice(index));
          index = raw.length;
        } else {
          commentParts.push(raw.slice(index, endIndex));
          index = endIndex + inBlock[1].length;
          inBlock = null;
        }
        continue;
      }

      if (inString) {
        const endIndex = findStringEnd(raw, index, inString[1]);
        if (endIndex < 0) {
          index = raw.length;
        } else {
          index = endIndex + inString[1].length;
          inString = null;
        }
        continue;
      }

      let matched = false;
      for (const marker of lineMarkers) {
        if (!raw.startsWith(marker, index)) continue;
        commentParts.push(raw.slice(index + marker.length));
        index = raw.length;
        matched = true;
        break;
      }
      if (matched) break;

      for (const block of blocks) {
        if (!raw.startsWith(block[0], index)) continue;
        const commentStart = index + block[0].length;
        const endIndex = raw.indexOf(block[1], commentStart);
        if (endIndex < 0) {
          commentParts.push(raw.slice(commentStart));
          index = raw.length;
          inBlock = block;
        } else {
          commentParts.push(raw.slice(commentStart, endIndex));
          index = endIndex + block[1].length;
        }
        matched = true;
        break;
      }
      if (matched) continue;

      for (const stringMarker of strings) {
        if (!raw.startsWith(stringMarker[0], index)) continue;
        index += stringMarker[0].length;
        inString = stringMarker;
        matched = true;
        break;
      }
      if (matched) continue;

      index += 1;
    }

    for (const commentPart of commentParts) {
      pushMatches(out, commentPart, lineNo);
    }
  }

  return out;
}

File: src-tauri/src/parsers/tag_scanner.rs
use super::languages::LangDef;
use once_cell::sync::Lazy;
use regex::Regex;

#[derive(Debug, Clone)]
pub struct FoundTag {
    pub kind: String,
    pub line_no: i64,
    pub text: String,
}

static TAG_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?i)\b(TODO|FIXME|HACK|NOTE|XXX)\b[ \t:\-]*([^\r\n]*)").unwrap()
});

fn push_matches(out: &mut Vec<FoundTag>, text: &str, line_no: i64) {
    if text.trim().is_empty() { return; }
    for caps in TAG_RE.captures_iter(text) {
        let kind = caps.get(1).unwrap().as_str().to_uppercase();
        let body = caps.get(2).map(|m| m.as_str().trim()).unwrap_or("");
        out.push(FoundTag {
            kind,
            line_no,
            text: body.chars().take(240).collect(),
        });
    }
}

pub fn scan_tags(content: &str, lang: Option<&LangDef>) -> Vec<FoundTag> {
    let Some(lang) = lang else { return vec![] };
    let lines: Vec<&str> = {
        let mut out = Vec::new();
        let mut start = 0usize;
        let b = content.as_bytes();
        let mut i = 0usize;
        while i < b.len() {
            if b[i] == b'\r' {
                out.push(&content[start..i]);
                if i + 1 < b.len() && b[i + 1] == b'\n' { i += 2; } else { i += 1; }
                start = i;
            } else if b[i] == b'\n' {
                out.push(&content[start..i]);
                i += 1; start = i;
            } else { i += 1; }
        }
        out.push(&content[start..]);
        out
    };
    let mut out = Vec::new();
    let mut in_block: Option<(&str, &str)> = None;
    for (line_index, raw) in lines.iter().enumerate() {
        let line_no = (line_index + 1) as i64;
        let mut index = 0usize;
        let mut in_string: Option<(&str, &str)> = None;
        let mut comment_parts: Vec<String> = Vec::new();
        let chars: Vec<char> = raw.chars().collect();
        while index < chars.len() {
            if let Some((_, end)) = in_block {
                let rest: String = chars[index..].iter().collect();
                if let Some(pos) = rest.find(end) {
                    comment_parts.push(rest[..pos].to_string());
                    index += pos + end.chars().count();
                    in_block = None;
                } else {
                    comment_parts.push(rest);
                    index = chars.len();
                }
                continue;
            }
            if let Some((_, end)) = in_string {
                let rest: String = chars[index..].iter().collect();
                if let Some(pos) = rest.find(end) {
                    index += pos + end.chars().count();
                    in_string = None;
                } else {
                    index = chars.len();
                }
                continue;
            }
            let rest: String = chars[index..].iter().collect();
            let mut matched = false;
            for m in lang.line {
                if rest.starts_with(m) {
                    comment_parts.push(rest[m.len()..].to_string());
                    index = chars.len();
                    matched = true;
                    break;
                }
            }
            if matched { break; }
            for &(start, end) in lang.block {
                if rest.starts_with(start) {
                    let after = index + start.chars().count();
                    let rest2: String = chars[after..].iter().collect();
                    if let Some(pos) = rest2.find(end) {
                        comment_parts.push(rest2[..pos].to_string());
                        index = after + pos + end.chars().count();
                    } else {
                        comment_parts.push(rest2);
                        in_block = Some((start, end));
                        index = chars.len();
                    }
                    matched = true;
                    break;
                }
            }
            if matched { continue; }
            for &(s, e) in lang.string {
                if rest.starts_with(s) {
                    index += s.chars().count();
                    in_string = Some((s, e));
                    matched = true;
                    break;
                }
            }
            if matched { continue; }
            index += 1;
        }
        for part in comment_parts {
            push_matches(&mut out, &part, line_no);
        }
    }
    out
}
